ghcide/src/Development/IDE/Plugin/HLS.hs

```diff
@@ -436,35 +437,55 @@ makeCompletions :: [(PluginId, CompletionProvider IdeState)]
-              where
-                  go acc [] = acc
-                  go (Completions (List ls)) (Completions (List ls2):rest)
-                      = go (Completions (List (ls <> ls2))) rest
-                  go (Completions (List ls)) (CompletionList (CompletionListType complete (List ls2)):rest)
-                      = go (CompletionList $ CompletionListType complete (List (ls <> ls2))) rest
-                  go (CompletionList (CompletionListType complete (List ls))) (CompletionList (CompletionListType complete2 (List ls2)):rest)
-                      = go (CompletionList $ CompletionListType (complete || complete2) (List (ls <> ls2))) rest
-                  go (CompletionList (CompletionListType complete (List ls))) (Completions (List ls2):rest)
-                      = go (CompletionList $ CompletionListType complete (List (ls <> ls2))) rest
-          makeAction (pid,p) = do
+          combine cs = go True mempty cs
+
+          go !comp acc [] =
+            CompletionList (CompletionListType comp (List $ DList.toList acc))
+          go comp acc (Completions (List ls) : rest) =
+            go comp (acc <> DList.fromList ls) rest
+          go comp acc (CompletionList (CompletionListType comp' (List ls)) : rest) =
+            go (comp && comp') (acc <> DList.fromList ls) rest
+
+          -- | Process a list of completion providers until we reach a max number of results
+          makeAction ::
+            Int ->
+            [(PluginId, CompletionProvider IdeState)] ->
+            IO [Either ResponseError CompletionResponseResult]
+          makeAction 0 _ = return []
+          makeAction _ [] = return []
+          makeAction n ((pid, p) : rest) = do
             pluginConfig <- getPluginConfig lf pid
-            if pluginEnabled pluginConfig plcCompletionOn
+            results <- if pluginEnabled pluginConfig plcCompletionOn
                then otTracedProvider pid "completions" $ p lf ideState params
                else return $ Right $ Completions $ List []
+            case results of
+              Right resp -> do
+                let (n', results') = consumeCompletionResponse n resp
+                (Right results' :) <$> makeAction n' rest
+              Left err ->
+                (Left err :) <$> makeAction n rest
 
       case mprefix of
           Nothing -> return $ Right $ Completions $ List []
           Just _prefix -> do
-            mhs <- mapM makeAction sps
+            mhs <- makeAction (maxCompletions config) sps
             case rights mhs of
                 [] -> return $ Left $ responseError $ T.pack $ show $ lefts mhs
                 hs -> return $ Right $ combine hs
 
+-- | Crops a completion response. Returns the final number of completions and the cropped response
+consumeCompletionResponse :: Int -> CompletionResponseResult -> (Int, CompletionResponseResult)
+consumeCompletionResponse n it@(CompletionList (CompletionListType _ (List xx))) =
```

Comment:
Rename `n` to limit? I usually think of `n` as the number of something, but here it isn't.